Add auto-date histogram aggregations documentation (#7630)
opensearch-project/documentation-website: _aggregations/bucket/auto-interval-date-histogram.md

    "properties" :  {
      "date_posted" : {
        "type" : "date",
        "format" : "yyyy-MM-dd"
      }
    }
  }
}
```
{% include copy-curl.html %}

Next, index the following documents into the `blogs` index:

```json
PUT blogs/_doc/1
{
  "name": "Semantic search in OpenSearch",
  "date_posted": "2022-04-17"
}
```
{% include copy-curl.html %}

```json
PUT blogs/_doc/2
{
  "name": "Sparse search in OpenSearch",
  "date_posted": "2022-05-02"
}
```
{% include copy-curl.html %}

```json
PUT blogs/_doc/3
{
  "name": "Distributed tracing with Data Prepper",
  "date_posted": "2022-04-25"
}
```
{% include copy-curl.html %}

```json
PUT blogs/_doc/4
{
  "name": "Observability in OpenSearch",
  "date_posted": "2023-03-23"
}

```
{% include copy-curl.html %}

To use the `auto_date_histogram` aggregation, specify the field containing the date or timestamp values. For example, to aggregate blog posts by `date_posted` into two buckets, send the following request:

```json
GET /blogs/_search
{
  "size": 0,
  "aggs": {
    "histogram": {
      "auto_date_histogram": {
        "field": "date_posted",
        "buckets": 2
      }
    }
  }
}
```
{% include copy-curl.html %}

> The response shows that the blog posts were aggregated into two buckets. The interval was automatically set to 1 year, with all three 2022 blog posts in one bucket and the 2023 blog post in another:

Reviewer: ```suggestion
The response shows that the blog posts were aggregated into two buckets. The interval was automatically set to 1 year, with all three 2022 blog posts collected in one bucket and the 2023 blog post in another:
```

> The response shows that the blog posts were aggregated into two buckets. The interval was automatically set to 1 year, with all three 2022 blog posts in one bucket and the 2023 blog post in another:

Reviewer: Added "collected"

```json
{
  "took": 20,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 4,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  },
  "aggregations": {
    "histogram": {
      "buckets": [
        {
          "key_as_string": "2022-01-01",
          "key": 1640995200000,
          "doc_count": 3
        },
        {
          "key_as_string": "2023-01-01",
          "key": 1672531200000,
          "doc_count": 1
        }
      ],
      "interval": "1y"
    }
  }
}
```

## Returned buckets

> Each bucket of the aggregation contains the following information:

Reviewer 2: "bucket of the aggregation" => "aggregation bucket"?

Reviewer: I can remove "of the aggregation" altogether.

> Each bucket of the aggregation contains the following information:

Reviewer: ```suggestion
Each bucket contains the following information:
```

```json
{
  "key_as_string": "2023-01-01",
  "key": 1672531200000,
  "doc_count": 1
}
```

> In OpenSearch, dates are internally stored as 64-bit integers representing timestamps in milliseconds since the epoch. In the aggregation response, each bucket `key` is returned as such a timestamp. The `key_as_string` value shows the same timestamp, but formatted as a date string based on the [`format`](#date-format) parameter. The `doc_count` field contains the number of documents in the bucket.

Reviewer 2: ```suggestion
In OpenSearch, dates are internally stored as 64-bit integers representing timestamps in milliseconds since the epoch. In the aggregation response, each bucket `key` is returned as such a timestamp. The `key_as_string` value shows the same timestamp but formatted as a date string based on the [`format`](#date-format) parameter. The `doc_count` field contains the number of documents in the bucket.
```

## Parameters

Auto-interval date histogram aggregations accept the following parameters.

Parameter | Data type | Description
:--- | :--- | :--- 
`field` | String | The field on which to aggregate. The field must contain the date or timestamp values. Either `field` or `script` is required.
`buckets` | Integer | The desired number of buckets. The returned number of buckets is less than or equal to the desired number. Optional. Default is `10`.

> `minimum_interval` | String | The minimum interval to be used. Specifying a minimum interval can make the aggregation process more efficient. Valid values are `year`, `month`, `day`, `hour`, `minute`, `second`. Optional.

Reviewer 2: ```suggestion
`minimum_interval` | String | The minimum interval to be used. Specifying a minimum interval can make the aggregation process more efficient. Valid values are `year`, `month`, `day`, `hour`, `minute`, and `second`. Optional.
```

> `time_zone` | String | Indicates to use a different time zone other than the default (UTC) for bucketing and rounding. You can specify the `time_zone` parameter as a [UTC offset](https://en.wikipedia.org/wiki/UTC_offset), such as `-04:00`, or an [IANA time zone ID](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones), such as `America/New_York`. Optional. Default is `UTC`. For more information, see [Time zone](#time-zone).

Reviewer 2: ```suggestion
`time_zone` | String | Specifies to use a time zone other than the default (UTC) for bucketing and rounding. You can specify the `time_zone` parameter as a [UTC offset](https://en.wikipedia.org/wiki/UTC_offset), such as `-04:00`, or an [IANA time zone ID](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones), such as `America/New_York`. Optional. Default is `UTC`. For more information, see [Time zone](#time-zone).
```

```json
{
  "key_as_string": "20230101T000000Z",
  "key": 1672531200000,
  "doc_count": 1
}
```

## Time zone

By default, dates are stored and processed in UTC. The `time_zone` parameter allows you to specify a different time zone for bucketing. You can specify the `time_zone` parameter as a [UTC offset](https://en.wikipedia.org/wiki/UTC_offset), such as `-04:00`, or an [IANA time zone ID](https://en.wikipedia.org/wiki/List_of_tz_database_time_zones), such as `America/New_York`.

As an example, index the following documents into an index:

```json
PUT blogs1/_doc/1
{
  "name": "Semantic search in OpenSearch",
  "date_posted": "2022-04-17T01:00:00.000Z"
}
```
{% include copy-curl.html %}

```json
PUT blogs1/_doc/2
{
  "name": "Sparse search in OpenSearch",
  "date_posted": "2022-04-17T04:00:00.000Z"
}
```
{% include copy-curl.html %}

First, run an aggregation without specifying a time zone: 

```json
GET /blogs1/_search
{
  "size": 0,
  "aggs": {
    "histogram": {
      "auto_date_histogram": {
        "field": "date_posted",
        "buckets": 2,
        "format": "yyyy-MM-dd HH:mm:ss"
      }
    }
  }
}
```
{% include copy-curl.html %}

> The response contains two 3-hour buckets, starting from midnight UTC on April 17, 2022:

Reviewer 2: ```suggestion
The response contains two 3-hour buckets, starting at midnight UTC on April 17, 2022:
```

```json
{
  "took": 6,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 2,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  },
  "aggregations": {
    "histogram": {
      "buckets": [
        {
          "key_as_string": "2022-04-17 01:00:00",
          "key": 1650157200000,
          "doc_count": 1
        },
        {
          "key_as_string": "2022-04-17 04:00:00",
          "key": 1650168000000,
          "doc_count": 1
        }
      ],
      "interval": "3h"
    }
  }
}
```

> Now, specify a `time_zone` of `-02:00` 

Reviewer 2: ```suggestion
Now, specify a `time_zone` of `-02:00`: 
```

```json
GET /blogs1/_search
{
  "size": 0,
  "aggs": {
    "histogram": {
      "auto_date_histogram": {
        "field": "date_posted",
        "buckets": 2,
        "format": "yyyy-MM-dd HH:mm:ss",
        "time_zone": "-02:00"
      }
    }
  }
}
```

> The response contains two buckets in which the start time is shifted by two hours and starts with 23:00 on April 16, 2022:

Reviewer 2: ```suggestion
The response contains two buckets in which the start time is shifted by 2 hours and starts at 23:00 on April 16, 2022:
```

```json
{
  "took": 17,
  "timed_out": false,
  "_shards": {
    "total": 1,
    "successful": 1,
    "skipped": 0,
    "failed": 0
  },
  "hits": {
    "total": {
      "value": 2,
      "relation": "eq"
    },
    "max_score": null,
    "hits": []
  },
  "aggregations": {
    "histogram": {
      "buckets": [
        {
          "key_as_string": "2022-04-16 23:00:00",
          "key": 1650157200000,
          "doc_count": 1
        },
        {
          "key_as_string": "2022-04-17 02:00:00",
          "key": 1650168000000,
          "doc_count": 1
        }
      ],
      "interval": "3h"
    }
  }
}
```

> When using time zones with daylight saving time (DST) changes, the size of buckets near the transition may differ slightly from the size of neighboring buckets. 

Reviewer 2: ```suggestion
When using time zones with daylight saving time (DST) changes, the size of a bucket that is near the transition may differ slightly from the size of a neighboring bucket. 
```

> {: .note}

Reviewer 2: Above: Does singular work here? Otherwise I think it needs to be "the sizes of buckets that are near the transition may differ slightly from the sizes of neighboring buckets."

---
layout: default
title: Auto-date histogram
parent: Bucket aggregations
grand_parent: Aggregations
nav_order: 15
---

> # Auto-date histogram

Author: @msfroh @bowenlan-amzn @sandeshkr419 Would one of you be available to technical review this PR? The doc team is closing content gaps for aggregations and need a technical review before we can published the doc. Please note that I used internal tools to assist in code example generation and Dev Tools to test those examples. Thank you.

> The `auto_date_histogram` aggregation automatically creates date histogram buckets based on the time range of your data. This aggregation is particularly useful when you are working with time-series data and want to visualize or analyze data over different time intervals without having to specify the bucket size manually.

Reviewer: ```suggestion
The `auto_date_histogram` aggregation automatically creates date histogram buckets based on the time range of your data. This aggregation is particularly useful when you are working with time-series data and want to visualize or analyze data over different time intervals without specifying the bucket size manually.
```

To use the `auto_date_histogram` aggregation, you need to specify the field containing the date or timestamp values. Additionally, you can provide optional parameters to customize the behavior of the aggregation.

> For example, you can use the `auto_date_histogram` aggregation to :

Author: ```suggestion
For example, you can use the `auto_date_histogram` aggregation to create date-based buckets:
```

> For example, you can use the `auto_date_histogram` aggregation to create date-based buckets:

Reviewer: ```suggestion
For example, you can use the following `auto_date_histogram` aggregation to create date-based buckets:
```

> ```json

Author: Technical reviewer: Please confirm this example is relevant to an OpenSearch user. I tested the example using Dev Tools. If another example is more appropriate, please replace the draft example with your example. Thank you.

GET /sample-index/_search
{
  "size": 0,
  "aggs": {
    "histogram": {
      "auto_date_histogram": {
        "field": "timestamp",
        "buckets": 2,

>         "minimum_interval": "day",

Reviewer: What does this parameter specify?

Author: @bowenlan-amzn Please help me address the information gap. See comment above.

Reviewer 3: The accepted units for minimum_interval are:

year
month
day
hour
minute
second

auto date histogram won't try the interval shorter than the one specified.

Author: @bowenlan-amzn Please confirm the revised text is technically accurate. Thank you.

>         "time_zone": "America/Los_Angeles",

Reviewer: What are the possible formats for `time_zone`?

Author: @bowenlan-amzn Please help me address the information gap. See comment above.

>         "format": "yyyy-MM-dd HH:mm:ss"

Reviewer: What are the possible formats for dates?

Author: @bowenlan-amzn Please help me address the information gap. See comment above.

Reviewer 3: For any time data (date field), format and timezone are consistent. I would expect we have a different documentation page about these

      }
    }
  }
}
```
{% include copy-curl.html %}

> #### Example response

Reviewer: Instead of this heading, I would use a sentence describing what the response contains.

Author: @bowenlan-amzn Please help me address the information gap. Can you explain the response from a novice user point of view?

Reviewer 3: If you can provide the data you have indexed, this can just be the response of the previous search request

Author: @bowenlan-amzn Please verify the revised text is technically accurate. Thank you.
